### ISSUE TITLE
Cannot convert png to ico larger than 255x255

### ISSUE THREAD

**Issue by maxloh** [ImageMagick/ImageMagick#1577]
### Prerequisites

- [o] I have written a descriptive issue title
- [o] I have verified that I am using the latest version of ImageMagick
- [o] I have searched [open](https://github.com/ImageMagick/ImageMagick/issues) and [closed](https://github.com/ImageMagick/ImageMagick/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

ImageMagick could only generate valid ico file smaller than or equal to 255x255
It fails to generate valid 256x256 ico that can be correctly display by Windows

### Steps to Reproduce

- Download `sample.png`
![sample](https://user-images.githubusercontent.com/35831069/58270795-326ff600-7dbd-11e9-99de-e88ffcaf2f89.png)
- Run `magick convert sample.png -define icon:auto-resize=255 255.ico`, the generated 255x255 ico file can be displayed by Windows (See attached 255.ico file)
- Run `magick convert sample.png -define icon:auto-resize=256 256.ico`, the generated 255x255 ico file cannot be displayed by Windows (See attached 256.ico file)

### System Configuration

- ImageMagick version: 

ImageMagick 7.0.8-46 Q16 x64 2019-05-19
- Environment (Operating system, version and so on): Windows 10 Pro 18362.116
- Additional information:

**Comment by fmw42** [comment]
This may not help, but in IM 7, use magick, not magick convert. The latter may use IM 6 code.

**Comment by maxloh** [comment]
@fmw42 I tried to use magick only but the file is invalid too.
```magick sample.png -define icon:auto-resize=256 256-no-convert.ico```
256-no-convert.ico https://drive.google.com/open?id=1bZ3j7Pvt-TeHwtiEaV0AovCT_fCbJPLq

**Comment by fmw42** [comment]
Yes, I thought it would not matter. But suggested it just in case. Sounds like a proper bug.

**Comment by urban-warrior** [comment]
Icons with a width or height of 256 are stored as PNG blobs.  Windows supports this but not all Windows clients can recognize this format.

**Comment by maxloh** [comment]
@urban-warrior Is it possible to store 256x256 ico as BMP blobs with ImageMagick, like what we got when we convert image to ico smaller than 256x256?

**Comment by heyakyra** [comment]
Not only is it a proper bug but it was supposed to be fixed in version 6.8.9-8 according to this forum thread back in 2014:
https://www.imagemagick.org/discourse-server/viewtopic.php?p=115232&sid=943a019ff62b1a5f60552ae4a1d9a034#p115232

> Icons with a width or height of 256 are stored as PNG blobs. Windows supports this but not all Windows clients can recognize this format.

Sounds worthy of a warning, not an error. Also seems related to #1505 as users may expect these blobs for smaller filesizes

I also noticed other places where users are having to work around this unfortunate behavior such as https://github.com/godotengine/godot/issues/21347#issuecomment-448690621

**Comment by dgonyier** [comment]
Related to this, I get icon files with a mix of PNG and BMP sizes in same icon file, which seems to confuse some applications (Windows Preview, etc.)

```bash
$ magick -version
Version: ImageMagick 7.0.10-27 Q16 i686 2021-02-07 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5)
Delegates (built-in): bzlib cairo djvu fftw fontconfig fpx freetype gslib gvc jbig jng jp2 jpeg lcms lqr lzma pangocairo png ps raqm raw rsvg tiff webp x xml zlib

$ magick convert source.png -alpha off -define icon:auto-resize="256,128,96,64,48,32,16"   testme.ico

$ magick identify testme.ico
testme.ico[0] PNG 256x256 256x256+0+0 8-bit sRGB 107757B 0.000u 0:00.004
testme.ico[1] ICO 128x128 128x128+0+0 8-bit sRGB 0.000u 0:00.004
testme.ico[2] ICO 96x96 96x96+0+0 8-bit sRGB 0.000u 0:00.002
testme.ico[3] ICO 64x64 64x64+0+0 8-bit sRGB 0.000u 0:00.001
testme.ico[4] ICO 48x48 48x48+0+0 8-bit sRGB 0.000u 0:00.001
testme.ico[5] ICO 32x32 32x32+0+0 8-bit sRGB 0.000u 0:00.001
testme.ico[6] ICO 16x16 16x16+0+0 8-bit sRGB 245523B 0.000u 0:00.001

# I discovered this "hack" by adding -compress jpeg, which does nothing
# but a side effect is that I get BMP formats for all sizes which
# Windows likes.
$ magick convert source.png -alpha off -define icon:auto-resize="256,128,96,64,48,32,16"  -compress jpeg  testme.ico

$ magick identify testme.ico
testme.ico[0] ICO 256x256 256x256+0+0 8-bit sRGB 0.016u 0:00.004
testme.ico[1] ICO 128x128 128x128+0+0 8-bit sRGB 0.000u 0:00.002
testme.ico[2] ICO 96x96 96x96+0+0 8-bit sRGB 0.000u 0:00.002
testme.ico[3] ICO 64x64 64x64+0+0 8-bit sRGB 0.000u 0:00.001
testme.ico[4] ICO 48x48 48x48+0+0 8-bit sRGB 0.000u 0:00.001
testme.ico[5] ICO 32x32 32x32+0+0 8-bit sRGB 0.000u 0:00.001
testme.ico[6] ICO 16x16 16x16+0+0 8-bit sRGB 408142B 0.000u 0:00.001

# Imagemagick really needs to create an icon file with all PNG or all BMP sizes,
# not a mix
# For now, I hope this "hack" doesn't get fixed, because it generates well-behaved ICO files...
```

**Comment by torkeldanielsson** [comment]
Omg, thank you so much for finding this "-compress jpeg" option!!!

It is INSANE that that is the solution, but this solved a multi-day rabbit-hole digging into why our icons would not show up.

I would call this a major bug - without the "-compress jpeg" application icons will not work in Windows 10/11 in at least the following locations:

 - task bar
 - desktop shortcut
 - start menu

**Comment by bonsaipropaganda** [comment]
> Not only is it a proper bug but it was supposed to be fixed in version 6.8.9-8 according to this forum thread back in 2014: https://www.imagemagick.org/discourse-server/viewtopic.php?p=115232&sid=943a019ff62b1a5f60552ae4a1d9a034#p115232
> 
> > Icons with a width or height of 256 are stored as PNG blobs. Windows supports this but not all Windows clients can recognize this format.
> 
> Sounds worthy of a warning, not an error. Also seems related to #1505 as users may expect these blobs for smaller filesizes
> 
> I also noticed other places where users are having to work around this unfortunate behavior such as [godotengine/godot#21347 (comment)](https://github.com/godotengine/godot/issues/21347#issuecomment-448690621)


So, have you heard of a fix to this yet? I can't seem to find a proper work around. I'm coming from Godot as well, and I can't seem to convert my png into a 256x256 ico file. It will convert to the other sizes but unfortunately it won't display correctly.

**Comment by Tachi107** [comment]
I have published a patch which should properly fix this issue: #8569

**Comment by STPR** [comment]
@dgonyier 
> $ magick convert source.png -alpha off -define icon:auto-resize="256,128,96,64,48,32,16"   testme.ico

Can you try inverting the size like this ?:

> $ magick convert source.png -alpha off -define icon:auto-resize="16,32,48,64,96,128,256" -compress zip "led-off.ico"

I don't understand why this has been pushed has an issue since it's supported by Windows...
It just makes the file much bigger to me and I've no issues with the previous behavior.

**Comment by fmw42** [comment]
On IM 7 use magick not magick convert. If I recall correctlye try 0 to get 256. I recall a comment about that but cannot confirm that. Search the forum to find that comment. If so then it is likely a bug that has not yet been fixed Sent from my iPhoneOn Feb 23, 2026, at 6:18 AM, STPR ***@***.***> wrote:﻿STPR left a comment (ImageMagick/ImageMagick#1577)
@dgonyier

$ magick convert source.png -alpha off -define icon:auto-resize="256,128,96,64,48,32,16"   testme.ico

Can you try inverting the size like this ?:

$ magick convert source.png -alpha off --define icon:auto-resize="16,32,48,64,96,128,256" -compress zip "led-off.ico"

I don't understand why this has been pushed has an issue since it's supported by Windows...
It just makes the file much bigger to me and I've no issues with the previous behaviour.

**Comment by STPR** [comment]
The problem with the patch of @Tachi107 is that it completly bypass the -compress parameter as written here:
[https://imagemagick.org/script/command-line-options.php#compress](https://imagemagick.org/script/command-line-options.php#compress)
`When writing an ICO file, you may request that the images be encoded in PNG format, by specifying Zip compression.`

**Comment by Tachi107** [comment]
That’s not what I changed. It was already the case for images of 255 pixels and below. To make the parameter work regardless the size, another patch would be needed.

**Comment by STPR** [comment]
A solution could be to NOT include PNG when the `-compress zip` parameter  is NOT specified.
And let the the sizes equal to 256 and above treated as PNG only when the `-compress zip` parameter is specified.

**Comment by STPR** [comment]
With your patch, contrary to what is usually done (compress the sizes equal to 256 or above), the 256x256 is now not converted to PNG.
Even if `-compress zip` is specified.
